### ISSUE TITLE
Make emitting an error necessary to acquire a `ErrorReported` token

### ISSUE THREAD

**Issue by varkor** [rust-lang/rust#69426]
`ErrorReported` indicates that a compiler error has been emitted, but currently it may be freely constructed. We should make it unconstructable, so that the only way to get one is to `.emit()` an error.

cc https://github.com/rust-lang/rust/pull/68434#discussion_r370961307 @eddyb @estebank

**Comment by mark-i-m** [comment]
> We could have it set a thread_local! that panics if not defused by some diagnostic emission on the same thread or something, but it might not be worth it if calls can be removed.

Regarding the unchecked constructor of `ErrorReported` in #93222...

**Comment by eddyb** [comment]
> We could have it set a `thread_local!` that panics if not defused by some diagnostic emission on the same thread or something, but it might not be worth it if calls can be removed.

I wrote that but now I believe it's unnecessary. Nothing I saw in that PR even *requires* the unchecked constructor, short of `references_error`.

**Comment by eddyb** [comment]
To make it more accessible, I'm replicating https://github.com/rust-lang/rust/pull/93222#discussion_r790217359 here:

> `quiet` is from https://github.com/rust-lang/rust/commit/5cf300d695c4ac6e4bdab8fe5c48de6b05b2cd96 from #84176.
> 
> @GuillaumeGomez do you know if that change can be reverted? It shouldn't have been merged, it makes it look like it's legitimate to disable diagnostics (to be clear, this is very dangerous, especially as more things rely on `delay_span_bug` and error emission).
> 
> There is no way to keep `rustdoc` doing what that PR does and still e.g. create `ty::Error`. An error *has to* be emitted to have `ty::Error`.
> 
> I don't think compromising the integrity of the rest of the system is worth getting access to method call annotations in code, which seems to be the only use of this misfeature.
> 
> AFAICT the only way to get the information you want is to run a *separate* compilation pass, and sinkhole errors at the actual point where they would be printed to `stderr`, while letting the compiler properly track the fact that errors happened, and then scavenge whatever data is available, despite those errors.

TL;DR: there is now a `quiet` flag inside the diagnostic machinery that silences *everything*, including load-bearing `delay_span_bug`s, and it's really easy to turn on.

Progress on `ErrorReported` cannot be made until that change is reverted.

**Comment by eddyb** [comment]
Random bikeshed for `ErrorReported`'s name if we take `delay_span_bug` into account: `Doomed`(`Compilation`).

**Comment by Manishearth** [comment]
I'm also looking at how it's used and ... that code should absolutely not be re-typechecking source, there should be better ways to get the information it needs. @GuillaumeGomez please talk to the compiler team and figure out how to best get the information you need there.

@eddyb The feature where this is used is unstable and behind a flag in large part due to the rustdoc team being quite wary of such a large change, so y'all should absolutely just rip out `quiet` in this PR, it's fine if the link-to-def feature is broken.

**Comment by GuillaumeGomez** [comment]
I opened a discussion on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/get.20method.20information.20without.20emitting.20errors). I'd really prefer to have an alternative (if there is one) before the PR is merged.
